### PR TITLE
initial implementation without tests and with edge case

### PR DESCRIPTION
**Pull request by juliogarciag**
@sir-dunxalot 

This is not meant to be merged yet, its just for a review.

I was trying to implement the exit-on-escape feature and i found two main issues right now that i want to discuss:
1. How to test it? The main problem is that we have to apply the event listener into the body document. I tried appling it to the modal element instead but that doesn't seem to work. We need to use body. The problem is that the ember helpers for integration doesn't seem to help here (they don't find by using `"body"` or  `App.rootElement`).
2. Because of the limitation of the body, there is an edge case i want to discuss: This will close all modals in the page at once, that is different from what the close button does. I will explain it with a drawing:

```
modal 1 = controller + view + template <-- escape key event (closes!)
modal N = controller + view + template <-- escape key event (closes!)
```

So, we should need a kind of collection of open modals. Every time a modal opens, it's added to the collection. Every time the modal closes, it's removed. When the escape key event, it verifies that the current view is the last modal on the collection and closes it. In this way, we always only closes the modal on top. What do you think of that?

### CODE REVIEW

File: addon/views/modal.js [hunk 1]
```diff
@@ -33,6 +34,23 @@ export default Em.View.extend(
     this.set('visible', true);
   },
 
+  /* Hide with Escape */
+  turnOnEscapeEvent: function() {
+    var _this = this;
+
+    this.set('keydownListener', function(event) {
```
Comment (sir-dunxalot) [inline addon/views/modal.js]: Why is this set as a property on the view here? It could just be a variable in this function or a property from the get go. For example:

``` js
export default Ember.View,extend({
  // ...

  onEsc: function(event) {
    if (event.which === _this.escapeKeyCode) {
      // ...
    }
  },

  turnOnEscapeEvent: function() {
    Em.$('body').on('keydown', this.get('onEsc'));
  }
});
```

File: addon/views/modal.js [hunk 2]
```diff
@@ -33,6 +34,23 @@ export default Em.View.extend(
...
+      if (event.which == _this.escapeKeyCode) {
+        _this.get('controller').send('closeModal', _this.get('outlet'));
+      }
+    });
+
+    Em.$('body').on('keydown', this.get('keydownListener'));
```
Comment (sir-dunxalot) [inline addon/views/modal.js]: To solve the problem of multiple modals, did you try using the view's [`keyDown` event documented here](http://emberjs.com/api/classes/Ember.View.html#toc_event-names)?

Maybe something like:

``` js
export default Ember.View,extend({
  // ...

  turnOnEscapeEvent: function(event) {
    // Close modal here
  }.on('keyDown')
});
```

File: addon/views/modal.js (hunk 1 above)
Comment (juliogarciag) [inline addon/views/modal.js]: Yes, it was the original implementation i used, but i was worried about the times the event will be registered. (every time the element is inserted)

File: addon/views/modal.js (hunk 2 above)
Comment (juliogarciag) [inline addon/views/modal.js]: I tried registering the event in the view but it never triggered unless i put an input there. It seems like normal elements (except from body) don't trigger keyboard events.

File: tests/integration/DOM-test.js [hunk 3]
```diff
@@ -184,3 +185,17 @@ test('Clicking modal overlay', function(assert) {
   });
 
 });
+
+test('Pressing escape', function(assert) {
```
Comment (sir-dunxalot) [inline tests/integration/DOM-test.js:189]: As for how to test this, I think the first step will be to see if there is a way to limit the keyDown event to the 'topmost' modal. To test, you'll probably use the [keyEvent helper](http://emberjs.com/guides/testing/test-helpers/#toc_asynchronous-helpers) and then:

```
import { testModal } from '../helpers/modal-tests';
```

Comment (sir-dunxalot) [inline tests/integration/DOM-test.js:189]: [Here is an example of how to use testModal](https://github.com/sir-dunxalot/ember-modals/blob/master/tests/integration/options-test.js#L152).

File: tests/integration/DOM-test.js [hunk 4]
```diff
@@ -184,3 +185,17 @@ test('Clicking modal overlay', function(assert) {
...
+
+  visit('/');
+
+  showModal(templateName);
+
+  // keyEvent('body', 'keydown', escapeKeyCode); // this doesn't work :(
```
Comment (sir-dunxalot) [inline tests/integration/DOM-test.js]: You'll probably need to run this inside an `andThen` helper:

``` js
test('Pressing escape', function(assert) {
  visit('/');

  andThen(function() {
    // Check stuff here
  });
});
```

**Comment by sir-dunxalot** [comment]
Hey Julio, thanks for the PR! It will be great to merge this in. I've added some comments to the code.

**Comment by juliogarciag** [comment]
Hi. Thanks for the comments. Actually, i don't want to merge it before solving both problems. (multiple modals and tests). I added some replies too.

**Comment by juliogarciag** [comment]
I will try to make the event to be triggered from the modal element rather than the body (i think giving it the focus in some way will be enough). If it works reliably, the code will be simplified a lot.

**Comment by juliogarciag** [comment]
I changed the overall strategy: Instead of applying the event to the body, i made the view to trigger keyboard evets. The trick was done with a combination of calling `this.$().focus()` and the tabindex attribute. There's still one edge case thought: When two modals are open and we close the element on the front, the element behind doesn't get the focus again and thus can't trigger the keydown event.

**Comment by sir-dunxalot** [comment]
Great, I like the new strategy much better.

As for the edge case, that's a tricky one. A possible solution might be to add a focus method that runs after `this.disconnectOutlet` runs in the application route mixin. Perhaps:

``` js
    removeModal: function(outlet, parentView) {
      this.disconnectOutlet({
        outlet: outlet,
        parentView: parentView
      });

      // Check if another modal exists then...
      Ember.$('[role="dialog"]').focus();
    }
```

**Comment by juliogarciag** [comment]
I got it working with this:

``` javascript
    removeModal: function(outlet, parentView) {
      this.disconnectOutlet({
        outlet: outlet,
        parentView: parentView
      });

      Ember.run.later(function() {
        var viewElement = Ember.$('[role="dialog"]').closest('.ember-view');
        viewElement.removeAttr('tabindex').attr('tabindex', 1);
        viewElement.focus();
      });
    }
```

I had to remove the tabindex and add it again to make the focus to work. It's a bit messy code and i think it should go on the view, but i don't know of a way to get the view instance of the reamining modal here, so i think this is enough for now. What do you think?

File: addon/mixins/routes/application.js [hunk 5]
```diff
@@ -31,6 +31,13 @@ export default Ember.Mixin.create({
         outlet: outlet,
         parentView: parentView
       });
+
+      // Check if another modal exists then give it focus
+      Ember.run.later(function() {
```
Comment (sir-dunxalot) [inline addon/mixins/routes/application.js]: What's the benefit of using `Ember.run.later` without a duration as the third argument? If it's absolutely necessary to use the run loop perhaps `Ember.run(function()` or `Ember.run.next(function()` would suffice.

File: addon/mixins/routes/application.js [hunk 6]
```diff
@@ -31,6 +31,13 @@ export default Ember.Mixin.create({
...
+        var remainingViewElement = Ember.$('[role="dialog"]:last-child').closest('.ember-view');
```
Comment (sir-dunxalot) [inline addon/mixins/routes/application.js:37]: Do you need to give the focus to the overlay or can you just give it to the modal dialog (e.g. not use `closest()`). The reason I ask is because `.closest('.ember-view')` seems a bit brittle.

Comment (sir-dunxalot) [inline addon/mixins/routes/application.js:37]: Also, does this code throw an error when there are no more modals rendered? I would expect `remainingViewElement` to not have a `removeAttr` method if there are no other modals present. It might be worth throwing in an `if` statement as a safety net here.

File: addon/mixins/routes/application.js (hunk 5 above)
Comment (juliogarciag) [inline addon/mixins/routes/application.js]: I didn't try the other options to run on the run loop. I'm not too familiar yet but when running outside of it, the `$` was returning also the element that was going to be removed before it was removed, so the calls to the jquery object returned by that were going to apply to both elements.

File: tests/integration/DOM-test.js [hunk 7]
```diff
@@ -184,3 +185,19 @@ test('Clicking modal overlay', function(assert) {
...
 });
+
+test('Pressing escape', function(assert) {
+
+  visit('/');
+
+  showModal(templateName);
+
+  andThen(function() {
+    keyEvent(inspect('overlay'), 'keydown', escapeKeyCode);
```
Comment (sir-dunxalot) [inline tests/integration/DOM-test.js:196]: keyEvent is a sync helper so you shouldn't need to use nested `andThen`s. I haven't used it before but try the following:

``` js
visit('/');

showModal(templateName);

keyEvent(inspect('overlay'), 'keydown', escapeKeyCode);

andThen(function() {
  assert.ok(/* ... */)
});    
```

Sorry if my first comment wasn't very clear.

File: addon/mixins/routes/application.js (hunk 6 above)
Comment (juliogarciag) [inline addon/mixins/routes/application.js:37]: Yes, it will work without an error because it's a jquery object and jquery objects are wrappers over an array of elements. Even if no element was found, it's still a wrapper over an empty array so the call will just pass through. For example, this works without errors everywhere:

``` javascript
$('div:first').closest('.this-parent-element-wont-never-exists').removeAttr('name').addClass('ok')
```

Comment (juliogarciag) [inline addon/mixins/routes/application.js:37]: And the `closest` was needed because the `tabindex` was attached to the ember view. I think i can just remove the `tabindex` attribute binding from the view and add it into the template to avoid using `closest`.

File: tests/integration/DOM-test.js (hunk 7 above)
Comment (juliogarciag) [inline tests/integration/DOM-test.js:196]: Yes, `keyEvent` is synchronous but i think `inspect` or `showModal` aren't because when calling in this way, the test rants about `inspect('overlay')` not being found.

File: addon/mixins/routes/application.js (hunk 6 above)
Comment (juliogarciag) [inline addon/mixins/routes/application.js:37]: But i tested without using the `closest` call and it seems like the focus is bubbling correctly so i will only remove the `closest` call.

Comment (juliogarciag) [inline addon/mixins/routes/application.js:37]: Without the `closest` call the modal will get the focus and we have to add some custom css to hide the border given to focused elements. Curiously, when the view itself has the focus, the border doesn't appear.

File: addon/mixins/routes/application.js (hunk 5 above)
Comment (sir-dunxalot) [inline addon/mixins/routes/application.js]: Ok. Try `Ember.run.next` - that will schedule the focus to take place in the next run loop (after the modal has been removed).

Comment (sir-dunxalot) [inline addon/mixins/routes/application.js]: When you use `Ember.run.later` it's intended to received a [third argument](http://emberjs.com/api/classes/Ember.run.html#method_later). I would say use run later as a last resort because it's kind of hacky here.

File: addon/mixins/routes/application.js (hunk 6 above)
Comment (sir-dunxalot) [inline addon/mixins/routes/application.js:37]: Ok. Well it sounds like it has advantages and benefits. I'm happy to go with whatever you decide.

File: tests/integration/DOM-test.js (hunk 7 above)
Comment (sir-dunxalot) [inline tests/integration/DOM-test.js:196]: hmm, that's very odd because `showModal` is an async helper too. If the test is working that's good enough for now.

**Comment by sir-dunxalot** [comment]
Overall this is looking very good. Let me know when you believe the PR is at the stage for final check and merge.

File: addon/mixins/routes/application.js (hunk 5 above)
Comment (juliogarciag) [inline addon/mixins/routes/application.js]: Thanks for the explanation of `Ember.run`. I am pretty new to ember and i didn't knew that. In my work we use angular (u_u) but i use ember in personal projects and freelance.

**Comment by juliogarciag** [comment]
I think i feel sufficiently confortable with it now.

File: addon/mixins/routes/application.js (hunk 5 above)
Comment (sir-dunxalot) [inline addon/mixins/routes/application.js]: No worries! The run loop is how Ember groups together executing its internal workings to improve performance. [This post](http://stackoverflow.com/questions/13597869/what-is-ember-runloop-and-how-does-it-work) explaining it in more detail is pretty legendary. `Ember.run.later` is equivalent in function to setTimeout - you use it to delay the execution of some code. A typical use case might be delaying a closing animation or showing the user something for a certain amount of time. It's a dangerous this to use to 'make sure the correct values are set' in cases like that you're better off using `Ember.run.next` (executes something in the next run loop) or observing a property. `Ember.run.sceheduleOnce` can be useful when you want to tun the function at a certain point in the run loop (e.g. `afterRender`, when DOM is ready).

**Comment by sir-dunxalot** [comment]
Merging and pulling down. Will release at 0.3.6 shortly.
